From cd17e889a9b162ce15c86c94ff4043f668172e4e Mon Sep 17 00:00:00 2001
From: Guoqing Jiang <guoqing.jiang@suse.com>
Date: Fri, 1 Apr 2022 10:40:30 +0800
Subject: [PATCH] Revert "x86/fpu: Prevent state corruption in
 __fpu__restore_sig()"
Patch-mainline: Never, revert stable patch which is added again in sorted section
References: bsn#178
Modified-by-SEL: No


This reverts commit 076f732b16a5bf842686e1b43ab6021a2d98233e.

Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/x86/kernel/fpu/signal.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kernel/fpu/signal.c b/arch/x86/kernel/fpu/signal.c
index d5bc96a536c2..a4ec65317a7f 100644
--- a/arch/x86/kernel/fpu/signal.c
+++ b/arch/x86/kernel/fpu/signal.c
@@ -405,7 +405,14 @@ static int __fpu__restore_sig(void __user *buf, void __user *buf_fx, int size)
 	if (use_xsave() && !fx_only) {
 		u64 init_bv = xfeatures_mask_user() & ~user_xfeatures;
 
-		ret = copy_user_to_xstate(&fpu->state.xsave, buf_fx);
+		if (using_compacted_format()) {
+			ret = copy_user_to_xstate(&fpu->state.xsave, buf_fx);
+		} else {
+			ret = __copy_from_user(&fpu->state.xsave, buf_fx, state_size);
+
+			if (!ret && state_size > offsetof(struct xregs_state, header))
+				ret = validate_user_xstate_header(&fpu->state.xsave.header);
+		}
 		if (ret)
 			goto err_out;
 
-- 
2.26.2

